fgpyo/fasta/sequence_dictionary.py

```diff
@@ -0,0 +1,441 @@
+
+    def __index__(self) -> int:
+        return self.index
+
+
+@dataclass(frozen=True, init=True)
+class SequenceDictionary(Mapping[Union[str, int], SequenceMetadata]):
+    """Contains an ordered collection of sequences.
+
+    Attributes:
+        infos: the ordered collection of sequence metadata
+    """
+
+    infos: List[SequenceMetadata]
+    _dict: Dict[str, SequenceMetadata] = field(init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        # Initialize a mapping from sequence name to the sequence metadata for all names
+        self_dict: Dict[str, SequenceMetadata] = {}
+        for index, info in enumerate(self.infos):
+            if info.index != index:
+                raise ValueError(
+                    "Infos must be given with index set correctly."
+                    + f"  See ${index}th with name: {info.name}"
+                )
+            for name in info.all_names:
+                if name in self_dict:
+                    raise ValueError(f"Found duplicate sequence name: {name}")
+                self_dict[name] = info
+        object.__setattr__(self, "_dict", self_dict)
+
+    def same_as(self, other: "SequenceDictionary") -> bool:
+        """Returns true if the sequences share a common reference name (including aliases), have
+        the same length, and the same MD5 if both have MD5s"""
+        if len(self) != len(other):
+            return False
+        return all(this.same_as(that) for this, that in zip(self.infos, other.infos))
+
+    def to_sam(
+        self,
+        extra_header: Optional[Dict[str, Any]] = None,
+    ) -> pysam.AlignmentHeader:
+        """Converts the sequence dictionary to a `pysam.AlignmentHeader`.
+
+        Args:
+            extra_header: a dictionary of extra values to add to the header, None otherwise.  See
+                          `:~pysam.AlignmentHeader` for more details.
+        """
+        header_dict: Dict[str, Any] = {
+            "HD": {"VN": "1.5"},
+            "SQ": [meta.to_sam() for meta in self.infos],
+        }
+        if extra_header is not None:
+            header_dict = {**header_dict, **extra_header}
+        return pysam.AlignmentHeader.from_dict(header_dict=header_dict)
+
+    @staticmethod
+    @overload
+    def from_sam(header: pysam.AlignmentHeader) -> "SequenceDictionary": ...
+
+    @staticmethod
+    @overload
+    def from_sam(header: List[Dict[str, Any]]) -> "SequenceDictionary": ...
+
+    @staticmethod
+    def from_sam(
+        header: Union[pysam.AlignmentHeader, List[Dict[str, Any]]],
+    ) -> "SequenceDictionary":
+        if isinstance(header, pysam.AlignmentHeader):
+            return SequenceDictionary.from_sam(header=header.to_dict()["SQ"])
+
+        infos: List[SequenceMetadata] = [
+            SequenceMetadata.from_sam(meta=meta, index=index) for index, meta in enumerate(header)
+        ]
+
+        return SequenceDictionary(infos=infos)
```

Comment:
Similarly I think this should probably work from `sam.header["SD"]`.

Reply:
I think it does.  The `isinstance` checks if it's a `pysam.AlignmentHeader`, and if not, it assumes it's a `header.to_dict()["SQ"]`...  I'll update the docs